File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        pairs = {'(': ')', '{': '}', '[': ']'}
        stack = []
        for c in s:
            if c in pairs:
                stack.append(c)
            elif not stack or pairs[stack.pop()] != c:
                return False
        return not stack

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> map = Map.of(
            ')', '(', 
            '}', '{',
            ']', '['
        );
        
        Deque<Character> stack = new ArrayDeque<>();
        for (char c : s.toCharArray()) {
            if (!map.containsKey(c)) {
                stack.add(c);
                continue;
            }

            // TODO:
            if (stack.isEmpty()) return false;
            if (!map.get(c).equals(stack.peekLast())) return false;

            stack.pollLast();
        }

        return stack.isEmpty();
    }
}